### ISSUE TITLE
Omitting target does not build all stages like command line does

### ISSUE THREAD

**Issue by jdforsythe** [docker/build-push-action#377]
### Troubleshooting

Before sumbitting a bug report please read the [Troubleshooting doc](https://github.com/docker/build-push-action/blob/master/TROUBLESHOOTING.md).

### Behaviour

Multistage Dockerfile. One stage ("validation") is a throwaway - installs development dependencies and runs static analysis. On the command line, `docker build` builds all stages if you don't pass the `--target` flag and so this serves to fail the build if any of the static analysis fails.

The validation stage is "throwaway" because it's not required in any way by the final `release` stage. This is done to ensure dev dependencies never make it to the final image.

#### Steps to reproduce this issue

1. Make a multi-stage Dockerfile where one stage is not required for the final stage
2. Build with this action
3. Observe how that stage is never run

#### Expected behaviour

This is not documented anywhere that I can find in this repo and is surprising that it doesn't just work like the CLI `docker build`. If I don't supply a target stage, _it should build everything_. Or at least I should be able to pass a flag / value for "target" that makes it build everything.

The only suggestions I can find are to run two separate jobs, one with the validation stage as the target and the other with the final stage as the target. This defeats the purpose of the stages sharing common ancestors. All the common stages will be built twice, which costs time and build minutes.

#### Actual behaviour

This action _does not build the "validation" stage_. It's only building the stages required for the final stage.

### Configuration

* Repository URL (if public): 
* Build URL (if public): 

Dockerfile:
```yml
ARG NPM_TOKEN

## base stage
FROM node:12.6-alpine AS base

WORKDIR /usr/src/app

COPY package.json yarn.lock ./

## deps stage
FROM base AS dependencies
ARG NPM_TOKEN

RUN \ 
  # ... install apk dependencies
  && echo "//registry.npmjs.org/:_authToken=${NPM_TOKEN}" > .npmrc

RUN \
  ## install production dependencies to a separate folder (for layer caching)
  yarn install --production --modules-folder production_nm_folder --frozen-lockfile \
  ## install all dev and production dependencies
  && yarn install \
  && rm -f .npmrc

## validation stage
FROM dependencies AS validation

COPY . .

RUN \
  yarn run lint \
  && yarn run test \
  && yarn run check-deps

## compile stage
FROM dependencies AS compile

COPY . .

RUN yarn run raw-build

## release stage
FROM base AS release

COPY --from=dependencies /usr/src/app/production_nm_folder ./node_modules
COPY --from=compile /usr/src/app/lib ./lib

EXPOSE 8080

CMD ["node", "lib/index.js"]
```

Action:
```yml
name: Build Docker Image

on:

  ## when a release is published (not drafts)
  release:
    types: [published]

  ## on all pull requests (create, update, synchronize)
  pull_request:

env:
  IMAGE_NAME: ${{ github.repository }}

jobs:

  ## build image for testing purposes
  test:
    runs-on: ubuntu-latest
    if: github.event_name == 'pull_request'

    steps:
      ## build image (https://github.com/docker/build-push-action)
      - name: Build Docker Image
        id: docker_build
        uses: docker/build-push-action@v2
        with:
          build-args: |
            NPM_TOKEN=${{ secrets.NPM_TOKEN }}
          push: false

  ## build and publish image
  publish:
    runs-on: ubuntu-latest
    if: github.event_name == 'release'

    steps:

      ## Generate tags (https://github.com/docker/metadata-action)
      - name: Docker Meta
        id: meta
        uses: docker/metadata-action@v3
        with:
          images: ${{ github.repository }}
          tags: |
            type=ref,event=branch
            type=ref,event=pr
            type=ref,event=tag

      ## login to DockerHub for publishing (https://github.com/docker/login-action)
      - name: Login to DockerHub
        uses: docker/login-action@v1
        with:
          username: ${{ secrets.DOCKERHUB_USERNAME }}
          password: ${{ secrets.DOCKERHUB_TOKEN }}

      ## build and push image (https://github.com/docker/build-push-action)
      - name: Build and push
        id: docker_build
        uses: docker/build-push-action@v2
        with:
          build-args: |
            NPM_TOKEN=${{ secrets.NPM_TOKEN }}
          push: true
          tags: ${{ steps.meta.outputs.tags }}
```

### Logs

> Download the [log file of your build](https://docs.github.com/en/actions/managing-workflow-runs/using-workflow-run-logs#downloading-logs) and [attach it](https://docs.github.com/en/github/managing-your-work-on-github/file-attachments-on-issues-and-pull-requests) to this issue.

**Comment by crazy-max** [comment]
@jdforsythe

> The validation stage is "throwaway" because it's not required

That's the expected behavior for me because the last stage `release` didn't define `validation` stage but agree the classic builder will run the `validation` stage if no target is defined:

```Dockerfile
FROM alpine AS base

FROM base AS dependencies
RUN echo "Hello world!"
COPY ./Dockerfile /Dockerfile

FROM dependencies AS validation
RUN echo "VALIDATION"

FROM dependencies AS compile
RUN echo "COMPILE" > /compile

FROM base AS release
COPY --from=dependencies /Dockerfile /app/Dockerfile
COPY --from=compile /compile /app/compile
```

```
$ docker buildx build --progress plain .
#1 [internal] load build definition from Dockerfile
#1 sha256:b17207ca462d68595cba0fe4becd2c25497186d856008c11eefd4d372868e019
#1 transferring dockerfile: 389B 0.0s done
#1 DONE 0.1s

#2 [internal] load .dockerignore
#2 sha256:3fb23b21cba0fb028e63e7c359a40eb46a82570aa50d7fd9137c255115854fc4
#2 transferring context: 47B 0.0s done
#2 DONE 0.1s

#3 [internal] load metadata for docker.io/library/alpine:latest
#3 sha256:d4fb25f5b5c00defc20ce26f2efc4e288de8834ed5aa59dff877b495ba88fda6
#3 DONE 1.3s

#10 [internal] load build context
#10 sha256:6283854e451fdb130b491c663051085ed7ba33e92543f2ca1fdee140319a452f
#10 transferring context: 2.42kB 0.0s done
#10 DONE 0.1s

#4 [base 1/1] FROM docker.io/library/alpine@sha256:234cb88d3020898631af0ccbbcca9a66ae7306ecd30c9720690858c1b007d2a0
#4 sha256:6df90944a00582804b68d88e0a62284a6b2253b2fa4a6677bac32b4b90f4a25f
#4 resolve docker.io/library/alpine@sha256:234cb88d3020898631af0ccbbcca9a66ae7306ecd30c9720690858c1b007d2a0 0.0s done
#4 sha256:234cb88d3020898631af0ccbbcca9a66ae7306ecd30c9720690858c1b007d2a0 1.64kB / 1.64kB done
#4 sha256:1775bebec23e1f3ce486989bfc9ff3c4e951690df84aa9f926497d82f2ffca9d 528B / 528B done
#4 sha256:d4ff818577bc193b309b355b02ebc9220427090057b54a59e73b79bdfe139b83 1.47kB / 1.47kB done
#4 DONE 0.1s

#5 [dependencies 1/2] RUN echo "Hello world!"
#5 sha256:2bcde5f9e1af65f759797b78faa3c4d5ef422f071ff4e7f8da614d60a281b76e
#5 0.318 Hello world!
#5 DONE 0.4s

#6 [dependencies 2/2] COPY ./Dockerfile /Dockerfile
#6 sha256:be1cb1c5929fcb4e26caebf90ed6b69ccabf32f590a3a67a13ad4ec4787b5aa7
#6 DONE 0.1s

#7 [release 1/2] COPY --from=dependencies /Dockerfile /app/Dockerfile
#7 sha256:d6b6aa95b60b7e8240fc46be1c015fb25f8f05c1dbfd14d484be9bd76be716c2
#7 DONE 0.1s

#8 [compile 1/1] RUN echo "COMPILE" > /compile
#8 sha256:ccfcdb28e673e9fda01c2f1968bf80869dbf70f4179a5c834f81998caa418a24
#8 DONE 0.4s

#9 [release 2/2] COPY --from=compile /compile /app/compile
#9 sha256:5cb76a89b418f08a280397c289f35a75e43bdd226bff10b9a519ea47933f501e
#9 DONE 0.1s

#11 exporting to image
#11 sha256:e8c613e07b0b7ff33893b694f7759a10d42e180f2b4dc349fb57dc6b71dcab00
#11 exporting layers
#11 exporting layers 0.1s done
#11 writing image sha256:00bd320ec57d8be10bba8fc6333d1a0abc735e92157036571873a09b6bdb882c done
#11 DONE 0.1s
```

```
$ DOCKER_BUILDKIT=0 docker build .
Sending build context to Docker daemon  1.437MB
Step 1/11 : FROM alpine AS base
latest: Pulling from library/alpine
5843afab3874: Already exists
Digest: sha256:234cb88d3020898631af0ccbbcca9a66ae7306ecd30c9720690858c1b007d2a0
Status: Downloaded newer image for alpine:latest
 ---> d4ff818577bc
Step 2/11 : FROM base AS dependencies
 ---> d4ff818577bc
Step 3/11 : RUN echo "Hello world!"
 ---> Running in 85041482996e
Hello world!
Removing intermediate container 85041482996e
 ---> 3b28c1f1e5da
Step 4/11 : COPY ./Dockerfile /Dockerfile
 ---> f52f314cc1a6
Step 5/11 : FROM dependencies AS validation
 ---> f52f314cc1a6
Step 6/11 : RUN echo "VALIDATION"
 ---> Running in de2e7da12ff6
VALIDATION
Removing intermediate container de2e7da12ff6
 ---> f955f3bef713
Step 7/11 : FROM dependencies AS compile
 ---> f52f314cc1a6
Step 8/11 : RUN echo "COMPILE" > /compile
 ---> Running in 07d51604a4bb
Removing intermediate container 07d51604a4bb
 ---> b3009ad5849a
Step 9/11 : FROM base AS release
 ---> d4ff818577bc
Step 10/11 : COPY --from=dependencies /Dockerfile /app/Dockerfile
 ---> 1716993f481a
Step 11/11 : COPY --from=compile /compile /app/compile
 ---> 30b558408f46
Successfully built 30b558408f46
```

Looks linked to moby/buildkit#532. WDYT @tonistiigi @thaJeztah?

**Comment by thaJeztah** [comment]
This is the expected behaviour with buildkit, and actually was a limitation of the classic (non-buildkit) builder.

BuildKit will only build stages that are needed for the final target; stages that are not needed to produce the final target are skipped, which allows "optional", or "target-dependent" paths in your Dockerfile.

For example:

```dockerfile
ARG ENABLE_VERIFY=false

FROM busybox AS base
RUN echo "do some things"
COPY some files /src

FROM base AS verify-false

FROM base AS verify-true
RUN echo "run verify steps"

FROM verify-${ENABLE_VERIFY} AS final
```

The above allows building with `--build-arg ENABLE_VERIFY=true` to enable the "verify" step, but defaults to skipping that stage.

Other uses as platform-specific stages (e.g. a combined Windows and Linux dockerfile, or stages that are only needed for a specific architecture).


For your specific case; for BuildKit to build a stage it must be a dependency of the final stage. For your specific case, I think there's two options;

Before:

```dockerfile
FROM dependencies AS validation
COPY . .
RUN \
  yarn run lint \
  && yarn run test \
  && yarn run check-deps

## compile stage
FROM dependencies AS compile
COPY . .
RUN yarn run raw-build
```

### Option 1: make your compile stage "extend" the validation stage

Copy the source code once (`COPY . . `), then run validation before compiling. The Dockerfile below still allows running _only_ the validation step, but when compiling the validation step _must_ be executed:

```dockerfile
FROM dependencies AS validation
COPY . .
RUN \
  yarn run lint \
  && yarn run test \
  && yarn run check-deps

## compile stage
FROM validation AS compile
RUN yarn run raw-build
```

I would also consider splitting the `RUN` step in the validation stage; the validation stage is an intermediate stage, so producing additional layers should not be a problem; even if those layers _would_ be part of the final image, chances are that they either do not produce overlapping files (so no size difference), or don't produce any files. Making the validation steps separate `RUN` commands makes the output of your docker build more descriptive (you can see _what_ failed in validation (lint, test, or check-deps);

```dockerfile
FROM dependencies AS validation
COPY . .
RUN yarn run lint
RUN yarn run test
RUN yarn run check-deps

## compile stage
FROM validation AS compile
RUN yarn run raw-build
```

### Option 2: run the stages in parallel

This approach takes better advantage of BuildKit allowing stages to be run _in parallel_, and also allows you to run `validation` and `compile` stages separately (which can be useful during development, where possibly you only want to compile, but want to skip validation while making code changes).

To run stages in parallel, you can use `COPY --from` to add a stage as "dependency", and to force it being built. BuildKit will now build those stages, but when constructing the "build tree", it can detect if steps can be executed in parallel. BuildKit currently requires "something" to be copied (in the `COPY --from`); possibly some enhancements can be made in future to make this optional, but for your validation stage, you could generate logs, or perhaps a small file that includes the validation results.

The Dockerfile below will:

- if no target is specified:
    - build the `release` stage
    - in parallel:
        - start building the `dependencies` stage
            - this triggers the `base` stage to be built (as dependency) first
        - start building the `validation` stage
            - this triggers the `source` -> `dependencies` -> `base` stages to be built first
        - start building the `compile` stage
            - this triggers the `source` -> `dependencies` -> `base` stages to be built first
    - as mentioned above, the `source` stage is a dependency for the `validation` and `compile stages`
    - the `dependencies` stage is also a dependency for the `source` stage (but it would already be built, because it is also specified it in `COPY --from`

This Dockerfile also allows you to;

- build only the `compile stage`, without validation (`docker build --target=compile`)
- only run the validation stage, without compiling (`docker build --target=validation`)
- run both; `docker build` without `--target`, or by specifying `--target=release`)

```dockerfile
FROM dependencies AS source
COPY . .

FROM source AS validation
COPY . .
RUN yarn run lint && echo "linting ok" >> /validation-results.log
RUN yarn run test && echo "test ok" >> /validation-results.log
RUN yarn run check-deps && echo "check-deps ok" >> /validation-results.log

## compile stage
FROM source AS compile
RUN yarn run raw-build


FROM base AS release

COPY --from=dependencies /usr/src/app/production_nm_folder ./node_modules
COPY --from=validation /validation-results.log /tmp/
COPY --from=compile /usr/src/app/lib ./lib
```

Looking at the above, the `COPY --from=dependencies` may be redundant; both the `validation` and `compile` stages already extend this stage, so it needs to complete before those are built. As an alternative, you can copy `/usr/src/app/production_nm_folder` from either the `validation` or `compile` stage. Copying it from the `validation` stage would mean that the "validation-results.log" is no longer needed;

```dockerfile
FROM dependencies AS source
COPY . .

FROM source AS validation
COPY . .
RUN yarn run lint
RUN yarn run test
RUN yarn run check-deps

## compile stage
FROM source AS compile
RUN yarn run raw-build


FROM base AS release

COPY --from=validation /usr/src/app/production_nm_folder ./node_modules
COPY --from=compile /usr/src/app/lib ./lib
```

Question for the above (I'm not familiar with your source code, so don't know what everytyhing does): the `validation` stage performs a `check-deps`: does this step only _validate_ or also (potentially) _update_ dependencies? If so, that step may be a requirement for the `compile` stage, otherwise the `compile` stage may be executed with different dependencies than are copied to the `release` stage.

**Comment by jdforsythe** [comment]
@thaJeztah 

The point of the separate `validation` stage and not copying anything from that is to ensure that no artifacts from those tools, like code coverage reports, ever make it into the final production image (`release` stage). This combined with the separate `yarn install` for only production dependencies and copying that to the `release` stage made it so the only thing that ends up in the final image are production dependencies and the compiled code.

To answer your question, the `check-deps` is just a security audit of dependencies (i.e. for node applications it runs `yarn audit --groups dependencies`). Nothing in the `validation` stage can make changes to anything that ends up in the final image. If anything in that stage fails, the developer needs to update the code in the repo to fix it.

I totally get why this feature can be useful, and the parallel stages is interesting. I will likely rewrite the Dockerfile to run the validation and release in parallel to save time. But both solutions you proposed require copying *something* from the validation stage in order to force it to run with a single build. I wouldn't want to copy a log over because this template is used across hundreds of repos and it's difficult to control what might end up in a log. I could just `touch` an empty file and copy it over, but that seems kinda hacky.

Is there any flag to get the build to behave like the CLI build does? Obviously `--target=all` wouldn't be good because a stage could be named "all", but perhaps a `--build-all-stages`? Or it might be nice to have some kind of command that specifically requires a previous stage, without using the `COPY --from`? Something like:

```yml
FROM alpine AS base

FROM base AS dependencies
COPY ./Dockerfile /Dockerfile

FROM dependencies AS validation
RUN echo "VALIDATION"

FROM dependencies AS compile
RUN echo "COMPILE" > /compile

## stage requires validation to execute but doesn't copy anything from that stage
FROM base AS release REQUIRES validation
COPY --from=dependencies /Dockerfile /app/Dockerfile
COPY --from=compile /compile /app/compile
```

I realize this should perhaps be a feature request in a separate repository. If you point me in the right direction, I would make it there.

**Comment by crazy-max** [comment]
@jdforsythe You can use [`buildx bake`](https://github.com/docker/buildx/blob/master/docs/reference/buildx_bake.md) for that purpose.
